### PR TITLE
Add subtitle/transcript capability

### PR DESCRIPTION
**Pull request by bondjimbond**
**JIRA Ticket**: https://jira.lyrasis.org/browse/ISLANDORA-2531

# What does this Pull Request do?

Allows Islandora Videojs to display subtitles if they exist on a given audio/video object.

# What's new?

If a datastream containing a WebVTT file exists on the object being viewed, if the DSID matches the Islandora Videojs configuration, the player will display them.

# How should this be tested?

* Check out this branch
* Add a new datastream to a video/audio object with the dsid TRANSCRIPT
* Upload a WebVTT file (doesn't have to match the actual video)
* View the object; subtitles should display.

Sample WebVTT file (save as `file.vtt`):

```
WEBVTT

00:00:00.500 --> 00:00:02.000
Here is some text

00:00:02.500 --> 00:00:04.300
and here is a bit more text.
```

# Additional Notes:
The meat of this code was originally written by @bryjbrown for use at FSU; I only adapted it to make it a general-use PR.

# Interested parties
@bryjbrown @whikloj @Islandora/7-x-1-x-committers

### CODE REVIEW

**Comment by Natkeeran** [comment]
@bondjimbond 
Does this require changes to the video content model?  I assume it will work without those changes; however, we may need to keep it consistent.
https://github.com/Islandora/islandora_solution_pack_video/blob/7.x/xml/islandora_video_ds_composite_model.xml

File: islandora_videojs.module
Comment (adam-vessey) [inline islandora_videojs.module]: Multisite/language path prefixing support:

```suggestion
    $variables['transcript_path'] = url("islandora/object/{$object->id}/datastream/{$transcript_dsid}/view");
```

File: islandora_videojs.module
Comment (jordandukart) [inline islandora_videojs.module:79]: Functionality wise this seems fine to my eyes however the [`ds-composite-model`](https://github.com/Islandora/islandora_solution_pack_video/blob/7.x/xml/islandora_video_ds_composite_model.xml) for the video SP doesn't define a `TRANSCRIPT` datastream and no references to transcript within the video SP. Not entirely sure what kind of precedent exists for something like this improvement at this time within Islandora 7.

File: theme/islandora-videojs.tpl.php
Comment (adam-vessey) [inline theme/islandora-videojs.tpl.php:19]: Is... there some form of escaping that should be done on the attribute contents here? Or... maybe back where they're being set in variables, in the `.module`? A `check_plain()` or something? If things somehow contained a `"`... things might go boom? And/or open up the door to adding other arbitrary markup?

File: islandora_videojs.module
Comment (bondjimbond) [reply islandora_videojs.module:79]: I would say a followup change to the video and audio SPs would be warranted, but not necessary -- this adds a capability to the viewer without actually affecting anyone who doesn't want to use it.

Comment (bondjimbond) [reply islandora_videojs.module:79]: Given how little development attention is paid to 7.x, I would not advise blocking a harmless improvement like this for such reasons.

File: theme/islandora-videojs.tpl.php
Comment (bondjimbond) [reply theme/islandora-videojs.tpl.php:19]: It seems to be escaped in videojs itself. I tried putting in quotes both closed and not closed, semicolons, and HTML tags both closed and unclosed. Seems to be fine. 

If you have specific concerns about markup that might be put into the VTT file I can test them.

File: islandora_videojs.module
Comment (bondjimbond) [reply islandora_videojs.module]: I have made this change.

**Comment by bondjimbond** [comment]
@Natkeeran It does not require changes, but it would be good to update the Video CModel to make TRANSCRIPT an optional datastream and provide for it. However, this PR should be merged first.

**Comment by bondjimbond** [comment]
@adam-vessey @Natkeeran  Given my responses above, does this meet standards for approval?

File: theme/islandora-videojs.tpl.php
Comment (bondjimbond) [reply theme/islandora-videojs.tpl.php:19]: Or if you meant the configured DSID? If you attempt to save a DS with invalid characters in the DSID, you get an error. 

So far as I can tell, putting invalid characters, markup, etc. into the variable in the form just results in it not finding the transcript datastream.

Comment (adam-vessey) [reply theme/islandora-videojs.tpl.php:19]: The label attribute? Especially given it's being scraped from the text? As in, if the "Language" line was something like:
```
Language: labeltext"><script>alert('bad things')</script><span consume="remaining
```

... might have to go further, to terminate the `<track/>` and `<video/>` elements to get it to actually run such a `<script/>` tag, but... is not insurmountable.

Comment (bondjimbond) [reply theme/islandora-videojs.tpl.php:19]: Aaaah, I see what you mean. I tried adding that language line to the vtt file, and did indeed get the alert.

I added a `check_plain()` to the language matching, and it seems to have worked. Does the implementation make sense?

Comment (adam-vessey) [reply theme/islandora-videojs.tpl.php:19]: Should do the trick, methinks.

**Comment by bondjimbond** [comment]
@jordandukart @Natkeeran I'll take a closer look at the Video CModel (and also Audio) after this is merged.

File: islandora_videojs.module
Comment (adam-vessey) [inline islandora_videojs.module]: Should probably additionally have an access check on the datastream (like an `islandora_datastream_access()` kind of thing; something of an example: https://github.com/Islandora/islandora_solution_pack_video/blob/085a8c3764b9ccf73bdfe6fc16eecd379badf171/theme/theme.inc#L59), before trying add something that will try to access it to the page.

File: islandora_videojs.module
Comment (adam-vessey) [inline islandora_videojs.module:78]: Something sits odd with me, about this... the [theme definition](https://github.com/Islandora/islandora_videojs/pull/53/files#diff-1c1ee2d3bd6211e965152b6e81e3f8df6eafbd96332ebe492907732c056d5651R47) above indicates that we allow/expect the object to be passed... but looking into things, it looks like it wouldn't be... though looks like it could easily be made to be passed in the [callback](https://github.com/Islandora/islandora_videojs/pull/53/files#diff-1c1ee2d3bd6211e965152b6e81e3f8df6eafbd96332ebe492907732c056d5651R95-R97) below... to make it something like:

```
function islandora_videojs_callback($params = array(), $object = NULL) {
  return theme('islandora_videojs', array('params' => $params, 'object' => $object));
}
```

As in, the [parameters for the callback](https://github.com/Islandora/islandora/blob/c34c7ac071ff0ac6802ea8701cb38a9f8ed49f76/includes/solution_packs.inc#L797) passes the object, our implementation here just doesn't touch it.

... if this change was made, we would then be able to access the object as `$variables['object']` here, instead of trying to do the load.

... that said, could live with this, as-is.

File: islandora_videojs.module
Comment (adam-vessey) [inline islandora_videojs.module:84]: If... there was indeed XACML preventing access to the datastream, I believe the attempted access here might lead to errors?... just reinforcing the need for the access check, I guess.

Comment (adam-vessey) [reply islandora_videojs.module:84]: ... also... half-wanting to suggest moving this to a `template_process_islandora_videojs()` implementation, given it's somewhat of a bigger bit of processing to load and regex in the transcript, but... yeah, probably not worth it at this point in time.

**Review by adam-vessey — CHANGES_REQUESTED** [review]
Really, just the access check stuff is _necessary_ to address, I think... the other bits would just be nice to have.

File: islandora_videojs.module
Comment (bondjimbond) [reply islandora_videojs.module]: Sorted

File: islandora_videojs.module
Comment (bondjimbond) [reply islandora_videojs.module:84]: Sorted

**Comment by bondjimbond** [comment]
@adam-vessey XACML stuff is sorted out. If you can live with the other bit, I'm happy to let you live with it, because I don't fully understand your suggestion.

File: islandora_videojs.module
Comment (bondjimbond) [reply islandora_videojs.module:78]: I think I'd prefer the "live with this" approach if it's OK with you.

**Review by adam-vessey — APPROVED** [review]

**Comment by bondjimbond** [comment]
@adam-vessey Ready to merge?

**Comment by adam-vessey** [comment]
@bondjimbond: In my mind, sure, but there was also input from others here... is it jumping the gun to merge without resolving it (or at least hearing from them)? Jordan's and Nat's comments?

**Comment by bondjimbond** [comment]
@adam-vessey I've responded to them with my comments, they never replied. If they chose not to reply, I think we're free to move forward.

**Comment by bondjimbond** [comment]
(Given there has been upward of 16 days to respond to my responses)

**Comment by DiegoPino** [comment]
Sorry for chiming in (trying to help a developer friend out here) but were not the rules 24 hours after an approval? Also, since this does not impose a condition on having the CMODEL XML definition changed, but uses DSIDs ,IF present (and CMODEL also being on another module) I personally feel approval is correct and this can be merged. I was not part of the code review but the pull is good. Thanks!

**Comment by bondjimbond** [comment]
@DiegoPino is correct - our rule, technically, is that after approval there are 24 hours open for comment before the PR can be merged. After 24 hours of no comments, community assent is presumed. 

Given that this much time that has passed since both the comments and responses as well as the approval, I don't see a reason to wait any longer.

**Comment by adam-vessey** [comment]
https://github.com/Islandora/islandora/wiki/Islandora-Committers-Workflow#merging-workflow , part 15: 

> If there are no outstanding concerns about the pull request, and the code has been sufficiently tested, and interested parties have been given sufficient time (24 hour window minimum) to respond to the pull request, then any Committer who feels confident with the code change (excluding those who would be considered ‘internal’ by the definition above) may merge the pull request.

... there were outstanding concerns, and I've not tested it (and don't presently have the capacity to do so). Perhaps I was wrong in adding my approval and not just dismissing my stale review?

**Comment by bondjimbond** [comment]
I saw no concerns with the code, besides yours, which were addressed -- the only questions outstanding are about other modules. This improvement doesn't actually touch those other modules - it only adds the capability to work with datastreams attached to objects if the DSID exists.

I believe we're fine to move forward on this with the goal of eventually making the TRANSCRIPT datastream "official" on those other modules -- but it's not required to do so right now, especially since the DSID here is configurable.

It is very hard to even generate discussion on improvements to I7 modules - questions about modules that are not materially affected by this change should not impact whether or not the improvement should happen.

**Comment by dannylamb** [comment]
Alright, let's pull this in.  It's been met with approval, albeit limited in some cases, and 24 hrs are way over.  I'll be the bad (good?) guy here and merge this.  If explosions occur, I'll gladly do the legwork to revert.

**Pull request merged**
